Implementing ByteStream interuption on infinate stream

simon-curtis opened the pull request:
# Description

This PR should address #13530 by explicitly handling ByteStreams. 

The issue can be replicated easily on linux by running:

```nushell
open /dev/urandom | into binary | bytes at ..10
```

Would leave the output hanging and with no way to cancel it, this was likely because it was trying to collect the input stream and would not complete.

I have also put in an error to say that using negative offsets for a bytestream without a length cannot be used.

```nushell
~/git/nushell> open /dev/urandom | into binary | bytes at (-1)..
Error: nu::shell::incorrect_value

  × Incorrect value.
   ╭─[entry #3:1:35]
 1 │ open /dev/urandom | into binary | bytes at (-1)..
   ·                                   ────┬─── ───┬──
   ·                                       │       ╰── encountered here
   ·                                       ╰── Negative range values cannot be used with streams that don't specify a length
   ╰────
   ```

# User-Facing Changes

No operation changes, only the warning you get back for negative offsets

# Tests + Formatting

Ran `toolkit check pr ` with no errors or warnings

Manual testing of the example commands above

WindSoilder commented on crates/nu-command/src/bytes/at.rs:
```diff
@@ -159,3 +146,88 @@ fn action(input: &Value, args: &Arguments, head: Span) -> Value {
+    engine_state: &EngineState,
+) -> Result<PipelineData, ShellError> {
+    let idxs = args.indexes;
+    if idxs.0 >= idxs.1 {
```
We should not make this early return.  For example: `0x[01 02 03] | bytes at 1..-1` should returns `0x[02 03]`.

simon-curtis replied:
Removed - added an example to cover the tests.

WindSoilder commented on crates/nu-command/src/bytes/at.rs:
Can you add some test cases for your change, I think it's good to test for the following:
```nushell
nu --testbin iecho 3 | bytes at 1..-3      # should returns an error
nu --testbin iecho 3 | bytes at 1..10     # should returns bytes
0x[01 02 03] | bytes at 1..-1     # should returns bytes
```

simon-curtis replied:
first two I have implemented as standalone tests - couldn't find a good place to put these tests so they're file scoped for the moment. The last example has been put in as an example and tests are passing

WindSoilder reviewed (changes requested):
Thank you!  The idea of using `ByteStream::from_fn` and `reader.bytes` looks good to me!

Just some comments for a corner case and some test cases

simon-curtis commented on crates/nu-cmd-lang/src/example_support.rs:
```diff
@@ -151,11 +152,19 @@ pub fn check_example_evaluates_to_expected_output(
+        let result = DebuggableValue(&result);
+        if result != expected {
+            panic!(
+                "\x1b[31mError:\x1b[0m The result of example \x1b[34m'{}'\x1b[0m for the command \x1b[34m'{}'\x1b[0m differs from the expected value.\n\n\
```
This error now looks like this:

```
Error: The result of example 'Extract bytes up to a negative index (inclusive)' for the command 'bytes at' differs from the expected value.

Expected: Binary([85, 16, 1, 19])
Actual:   Binary([16, 1, 19])
```

WindSoilder commented on crates/nu-command/src/bytes/at.rs:
```diff
@@ -159,3 +156,116 @@ fn action(input: &Value, args: &Arguments, head: Span) -> Value {
+}
+
+#[cfg(test)]
+mod tests {
```
Normally we put `nu!` relative tests into `nu-command/tests`.  For example, this can be put under `nu-command/tests/commands/bytes/at.rs`

simon-curtis replied:
I have put them in `tests/commands/bytes/at.rs`

WindSoilder commented on crates/nu-command/src/bytes/at.rs:
```diff
@@ -159,3 +156,116 @@ fn action(input: &Value, args: &Arguments, head: Span) -> Value {
+    use nu_test_support::nu;
+
+    #[test]
+    fn test_examples() {
```
You don't need to test example.  Nushell tests will test example automatically

simon-curtis replied:
Removed

WindSoilder reviewed (changes requested):
Thank you for the update!  I left some comments about the position of tests.  And CI is failed, you need to run `cargo fmt` to fix it :)

simon-curtis commented:
All requested changes have been completed 😄

IanManske commented on crates/nu-cmd-lang/src/example_support.rs:
```diff
@@ -159,11 +160,19 @@ pub fn check_example_evaluates_to_expected_output(
-        )
+        let expected = DebuggableValue(expected);
+        let result = DebuggableValue(&result);
+        if result != expected {
+            panic!(
+                "\x1b[31mError:\x1b[0m The result of example \x1b[34m'{}'\x1b[0m for the command \x1b[34m'{}'\x1b[0m differs from the expected value.\n\n\
+                Expected: {:?}\n\
+                Actual:   {:?}\n",
+                example.description,
+                cmd_name,
+                expected,
+                result
+            );
+        }
```
Thanks for the addded formatting! But I think `assert_eq` also allows a format string:

```suggestion
        assert_eq!(
            result,
            expected,
            "\x1b[31mError:\x1b[0m The result of example \x1b[34m'{}'\x1b[0m for the command\x1b[34m'{}'\x1b[0m differs from the expected value.\n\n\
Expected: {:?}\n
Actual:   {:?}\n",
                example.description,
                cmd_name,
                expected,
                result,
            );
        }
```

Hopefully, that formatting looks right?

simon-curtis replied:
For some reason it doesn't format the empty space the same as panic - so I have put it all on one line now

IanManske commented on tests/repl/tests.rs:
```diff
@@ -25,7 +25,7 @@ pub fn run_test(input: &str, expected: &str) -> TestResult {
     let name = file.path();
 
-    let mut cmd = Command::cargo_bin("nu")?;
+    let mut cmd: Command = Command::cargo_bin("nu")?;
```
Is this working around a type inference issue?

simon-curtis replied:
Not meant to be there - sorry. Im not sure why it was added, I will remove it.

IanManske reviewed (commented):
Thanks for this PR! The `take` and `skip` commands already support binary streams, and their code could be reused for this command. I.e., `bytes at ..$n` is essentially `take $n`, `bytes at $n..` is `skip $n`, and `bytes at $n..$m` is `skip $n | take ($m - $n)`. ~~I think it would make more sense if the binary stream handling was extracted out from those commands into a separate file and then reused across all these commands (`skip`, `take`, and `bytes at`).~~

simon-curtis commented:
I have just pushed a fix for a performance bug where in the "relative index" branch would collect the entire `ByteStream` before limiting.

One thing I have noticed is bytes from a File doesn't have a length?

![image](https://github.com/user-attachments/assets/0624234b-33f4-4e7a-bd33-0b607de7c29b)

IanManske commented:
Sorry, I should I have linked the code I was talking about. The `take` and `skip` commands I referred to before are other Nushell commands, not the `take` and `skip` defined on iterators in Rust.

https://github.com/nushell/nushell/blob/main/crates/nu-command/src/filters/skip/skip_.rs#L98C29-L114
https://github.com/nushell/nushell/blob/main/crates/nu-command/src/filters/take/take_.rs#L92-L105

You can reuse this code to implement `bytes at`.

simon-curtis commented:
Dw I got what you meant, I just wanted to tidy up the code first

simon-curtis reviewed (commented):
All changes have been made. I have merged the logic for `filters::skip` and `filters::take` into the ByteStream impl along with a `ByteStream::slice` function - which is now used in `bytes::at`. I've also added some more tests for regression

IanManske reviewed (commented):
Thanks for all the tests and refactoring! A few comments:

simon-curtis commented on crates/nu-command/src/bytes/at.rs:
```diff
@@ -159,3 +175,31 @@ fn action(input: &Value, args: &Arguments, head: Span) -> Value {
+    };
+
+    let end = match range.1 {
+        end if end < 0 => (len as isize + end) as usize,
```
Who's to say they don't want to select `18_446_744_073_709_551_615-n` items 😅 I can change it to a u32 maybe?

simon-curtis replied:
Yep. Bounds checks are needed. I will add them

simon-curtis commented on crates/nu-cmd-lang/src/example_support.rs:
```diff
@@ -159,11 +160,19 @@ pub fn check_example_evaluates_to_expected_output(
+            assert_eq!(
+                result,
+                expected,
+                "\x1b[31mError:\x1b[0m The result of example \x1b[34m'{}'\x1b[0m for the command \x1b[34m'{}'\x1b[0m differs from the expected value.\n\nExpected: {:?}\nActual:   {:?}\n",
```
Done

simon-curtis commented:
Thanks for the feedback btw, I'm learning a lot

IanManske commented on crates/nu-command/src/bytes/at.rs:
```diff
@@ -159,3 +174,28 @@ fn action(input: &Value, args: &Arguments, head: Span) -> Value {
     }
 }
+
+fn resolve_relative_range(range: &Subbytes, len: &usize) -> (usize, usize) {
```
No need to take a reference to a `usize`.
```suggestion
fn resolve_relative_range(range: &Subbytes, len: usize) -> (usize, usize) {
```

simon-curtis replied:
Thanks! After looking up why, I can definitely see why it would be less efficient here to use a ref instead of copy! 😊

IanManske commented on crates/nu-command/src/bytes/at.rs:
```diff
@@ -159,3 +174,28 @@ fn action(input: &Value, args: &Arguments, head: Span) -> Value {
+
+fn resolve_relative_range(range: &Subbytes, len: &usize) -> (usize, usize) {
+    let start = match range.0 {
+        start if start < 0 => len.checked_sub(start.unsigned_abs()).unwrap_or(0),
```
```suggestion
        start if start < 0 => len.saturating_sub(start.unsigned_abs()),
```

IanManske commented on crates/nu-command/src/bytes/at.rs:
```diff
@@ -159,3 +174,28 @@ fn action(input: &Value, args: &Arguments, head: Span) -> Value {
+    };
+
+    let end = match range.1 {
+        end if end < 0 => len.checked_sub(end.unsigned_abs()).unwrap_or(0),
```
```suggestion
        end if end < 0 => len.saturating_sub(end.unsigned_abs()),
```

IanManske reviewed (commented):
Thanks for the changes! This is almost good to go, just a few last things.

IanManske commented on crates/nu-protocol/src/pipeline/byte_stream.rs:
```diff
@@ -211,6 +211,81 @@ impl ByteStream {
+
+                self.skip(val_span, absolute_start as u64)
+                    .and_then(|stream| {
+                        let absolute_end = match end {
+                            end if end < 0 => (len as isize + end).max(0) as usize,
+                            end => end.min(len as isize) as usize,
+                        };
```
Same here.

IanManske commented on crates/nu-protocol/src/pipeline/byte_stream.rs:
```diff
@@ -211,6 +211,81 @@ impl ByteStream {
+                        if absolute_end < absolute_start {
+                            stream.take(val_span, 0)
+                        } else {
+                            stream.take(val_span, (absolute_end - absolute_start) as u64)
```
Does this need a `+ 1`? I saw that `map_value` does add 1.

simon-curtis commented:
I have moved logic into using the core IntRange instead of SubBytes as not to duplicate logic. 

I have added a couple of functions on the Range to resolve the absolute range values where a length is known too, returning the original Bound<i64> value to allow the caller to apply the index as they wish.

In our case we both apply it by slicing Vec<u8> and using the new ByteStream::slice function.

I am still testing and need to format so hold out for a min

simon-curtis commented:
I will fast forward and resolve merge conflicts tonight after work. AFAIK everything works and would be ready to merge in

WindSoilder commented:
Hi, @IanManske .  Can you take the last review when you get time?

IanManske reviewed (approved):
Thanks! This looks good at a glance, we'll land this to start testing in the wild.
